prover/crypto/state-management/smt/tree.go

```diff
@@ -282,26 +285,87 @@ func BuildComplete(leaves []types.Bytes32, hashFunc func() hashtypes.Hasher) *Tr
 	tree := NewEmptyTree(config)
 	tree.OccupiedLeaves = leaves
 
-	// Builds the tree bottom-up
+	// pre-allocate the memory for all levels
+	nbTotalNodes := 0
+	for d := 1; d < depth; d++ {
+		nbTotalNodes += 1 << (depth - d)
+	}
+	arena := make([]types.Bytes32, nbTotalNodes)
+	offset := 0
+
+	// start a worker pool that will parallelize the hashing at each level
+	type workerTask struct {
+		curr  []types.Bytes32
+		next  []types.Bytes32
+		start int
+		stop  int
+		wg    *sync.WaitGroup
+	}
+
+	workerCount := runtime.NumCPU()
+	if workerCount < 1 {
+		workerCount = 1
+	}
+
+	tasks := make(chan workerTask)
+	var workersWG sync.WaitGroup
+	workersWG.Add(workerCount)
+	for i := 0; i < workerCount; i++ {
+		go func() {
+			hasher := config.HashFunc()
+			for task := range tasks {
+				if task.stop > task.start {
```

Comment:
Duplicate check, in the loop below we always start from `k := task.start` until `k < task.stop`. If `task.stop < task.start` then we don't loop.